posthog/api/signup.py

```diff
@@ -44,16 +46,32 @@ def validate_password(self, value):
         return value
 
     def create(self, validated_data, **kwargs):
-        is_instance_first_user: bool = not User.objects.exists()
+        already_existing_demo_user = (
+            None if not settings.DEMO else User.objects.filter(email=validated_data["email"]).first()
+        )
 
-        organization_name = validated_data.pop("organization_name", validated_data["first_name"])
+        if not already_existing_demo_user:
+            # Normal signup flow
+            is_instance_first_user: bool = not User.objects.exists()
+
+            organization_name = validated_data.pop("organization_name", validated_data["first_name"])
+            if settings.DEMO:
+                validated_data["password"] = None
+
+            self._organization, self._team, self._user = User.objects.bootstrap(
+                organization_name=organization_name,
+                create_team=self.create_team,
+                **validated_data,
+                is_staff=is_instance_first_user,
+            )
+        else:
+            # If there's an email collision in signup in the demo environment, we treat it as a login
+            is_instance_first_user = False
+
+            self._user = already_existing_demo_user
+            self._organization = already_existing_demo_user.organization
+            self._team = already_existing_demo_user.team
```

Comment:
A nit perhaps, but could this be refactored to:

```
if settings.DEMO:
    # override everything
else:
    # normal flow
```

It's kind of hard to follow along, and such critical code as the signup flow should be made as mistake-proof as possible. Maybe already just flipping the if `if not already_existing_demo_user:` to `if already_existing_demo_user:` will make this cleaner?

Reply:
Agreed, refactored the demo handling to method `enter_demo()`.